### PR TITLE
test(twap): `updateRecords` function (two-asset)

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #2659

## What is the purpose of the change

Covers `updateRecords` function by tests

## Brief Changelog
- added tests
- refactored amm mock
    * Reason: It was impossible to set it up in such a way as to test the case with invalid number of records.
- Moved two errors to `types` - one in `gamm`, one in `twap`. This is to be able to make sure that the desired error is returned in tests.

This change is state-breaking relative to the previous RCs.

## Follow-Ups
- https://github.com/osmosis-labs/osmosis/issues/2686
- https://github.com/osmosis-labs/osmosis/issues/2688

## Testing and Verifying

This change added tests.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/twap/logic.go
Comment (p0mvn) [inline x/twap/logic.go:50]: Note to reviewers: please see "Changelog"  under PR description

Comment (p0mvn) [reply x/twap/logic.go:50]: Also, the above TODO makes sense IMO. Should we consider switching?

File: x/twap/logic.go
Comment (p0mvn) [inline x/twap/logic.go:55]: Note to reviewers: please see "Changelog" under PR description

Also, the TODO makes sense. Should we consider switching?

**Comment by p0mvn** [comment]
md linter seems to be broken on `main`

cc: @nicolaslara @AlpinYukseloglu might be related to #2409 or #2631

**Comment by AlpinYukseloglu** [comment]
> md linter seems to be broken on `main`
> 
> 
> 
> cc: @nicolaslara @AlpinYukseloglu might be related to #2409 or #2631

I was getting issues with the general linter as well but `make markdown` should resolve any markdown-related issues (& flag any it couldn't solve)

**Comment by nicolaslara** [comment]
I saw those issues on a branch after merging main. I'm guessing there was some change on what the linter checks. The rate limiter readme just needs a few blank lines around titles.

**Comment by nicolaslara** [comment]
@AlpinYukseloglu Added the readme changes here: https://github.com/osmosis-labs/osmosis/pull/2691

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:534]:
```suggestion
// - spot price error times are either propagated from
```

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:537]:
```suggestion
```

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:554]: nit: I think this could be a suite method + can be used in different places around our test code

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:536]: I think we have duplicated boundaries of testing with `TestUpdateTwap` with this test case that could be merged into one, wdyt?

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:712]: oh interesting so we cap at max spot price regardless of having greater value than sp

File: x/twap/logic_test.go
Comment (mattverse) [inline x/twap/logic_test.go:900]: let's create an issue for this so that we come back to this!

**Review by mattverse — APPROVED** [review]
LGTM!

File: x/twap/logic_test.go
Comment (p0mvn) [reply x/twap/logic_test.go:553]: I've given our tests a quick skim and couldn't find where this function and `expectedResults` type could be reused.

The problem here is that `expectedResults` is a type that is local to this test. I think if anyone decides to make that type global / apply it in other tests, then it would make sense.

File: x/twap/logic_test.go
Comment (p0mvn) [reply x/twap/logic_test.go:536]: `TestUpdateTwap` tests `updateRecord` method while this is testing `updateRecords` (plural).

I've referenced that test while implementing it, and, unfortunately, I could not find any logic suitable for reuse or de-duplication.

File: x/twap/logic_test.go
Comment (p0mvn) [reply x/twap/logic_test.go:711]: Yes we cap it if it is greater but we set it to zero if we get an error

File: x/twap/logic_test.go
Comment (p0mvn) [reply x/twap/logic_test.go:899]: Already made: https://github.com/osmosis-labs/osmosis/issues/2688

**Comment by p0mvn** [comment]
Comments addressed and waiting for one more approval

File: x/twap/logic_test.go
Comment (mattverse) [reply x/twap/logic_test.go:536]: Ah didn't notice the difference in test being plural -and singular. IMO we should change `TestUpdateTwap` -> `TestUpdateRecords` so that we can be consistent with the testing names

File: x/twap/logic_test.go
Comment (czarcas7ic) [inline x/twap/logic_test.go:605]: Are spaces allowed in denoms? If not, we should use a valid denom for the extra denom

File: x/twap/logic_test.go
Comment (czarcas7ic) [inline x/twap/logic_test.go:665]: Clarifying question: The description says "updated with spot price error in both denom pairs", there is a spErr in both denoms pairs, but we dont overrideSp for baseQuoteAB. Is this intentional? In my mind, if there is a spot price error in both denom pairs, then they should both return zero (as it currently stands, I am in 100 percent agreement with your recommendation that they should actually return the last known spot price)

File: x/twap/logic_test.go
Comment (czarcas7ic) [inline x/twap/logic_test.go:689]: nit: Maybe instead of "large spot price" we should specify "invalidly large spot price"

File: x/twap/logic_test.go
Comment (czarcas7ic) [inline x/twap/logic_test.go:927]:
```suggestion
				s.Require().ErrorIs(err, tc.expectError)
```
nit

**Review by czarcas7ic — APPROVED** [review]
This LGTM! 👍

**Comment by ValarDragon** [comment]
Something feels like it's gone quite wrong in testing design Here. 

This function were supposed to be testing is quite small and has minimal functionality yet has 400+ lines of tests

**Review by ValarDragon — APPROVED** [review]

File: x/twap/logic_test.go
Comment (p0mvn) [reply x/twap/logic_test.go:689]: It is not invalid to have a large spot price. However, we do cap it at a specific maximum

File: x/twap/logic_test.go
Comment (p0mvn) [reply x/twap/logic_test.go:665]: This is intentional as it currently stands.

We only set the spot price to zero if there is a spot price error and the returned value from amm is an empty `Dec`:
https://github.com/osmosis-labs/osmosis/blob/bcda1cbe01f87111eb9320a8907b6b89d9930fb6/x/twap/logic.go#L49-L51

I'm planning to change this logic in:
https://github.com/osmosis-labs/osmosis/issues/2689

**Comment by p0mvn** [comment]
> 

I noticed that it is possible to refactor and infer most recent records from the boolean flag in historical records:
[infer most recent from historical](https://github.com/osmosis-labs/osmosis/pull/2687/commits/68ede334686fcfe0aaffbe3788d2ad99587c110d)

I'm not seeing anything else that can be changed or removed. However, I'm open to suggestions.

This function does quite a bit:
- there are 2 edge cases (error branches).
- it mutates 2 database indexes. As a result, we need to test if they are updated correctly.
- there are a few extra edge cases stemming from the spot prices logic that I attempted to cover.

Also, the mock setup logic and overrides take up some space

**Comment by p0mvn** [comment]
Merging this, we can discuss further improving this test in a follow-up PR

**Comment by ValarDragon** [comment]
Yeah, focusing on reducing these lines of code is not at all important! But would like to figure out if we can make strategies to reduce this for similar cases in the future

**Pull request merged**
